feat: add leave button and notify other players

eduardhuck opened the pull request:
Open Issues:

- Uncaught TypeError: payload is undefined on leave room
- Uncaught TypeError: room is undefined on leave room
- Backend does not yet close rooms without players (new issue?)


@fafeitsch: Maybe you have an idea whats the origin of the TypeErrors.

eduardhuck committed:
feat: add leave button and notify other players

fafeitsch commented on frontend/src/app/map/Map.svelte:
```diff
@@ -91,6 +105,15 @@ function advanceGame() {
+    console.log('leave game');
+    window.location = window.location.protocol + '//' + window.location.host;
+    // TODO this seems to cause an uncaught typeError: payload is undefined
+    store.set.resetGame();
```
Your assumption is correct. The errors are thrown in Leaflet.svelte file.
1. In line 66 the component subscribes to the store's room in order to build the Leaflet Map. Up to now, the component did non exist while the room was undefined, thus, there is no undefined check. Either a `filter(room => !!room)` or a `take(1)` should do the trick, because the component is created freshly for every game.
2. In line 74, the component subscribes to the roomUpdated/roomJoined topic, to set up the map options correctly. However, in sockets.ts:21 an `undefined` is sent everytime the room is reset – the Leaflet component doesn't handle that case correctly yet. You could duplicate the filter from line 86 and move it to 79. But maybe there's a better solution.

eduardhuck replied:
>     2. In line 74, the component subscribes to the roomUpdated/roomJoined topic, to set up the map options correctly. However, in sockets.ts:21 an `undefined` is sent everytime the room is reset – the Leaflet component doesn't handle that case correctly yet. You could duplicate the filter from line 86 and move it to 79. But maybe there's a better solution.

The suggested solution works fine for now.

If I understand sockets.ts:21 correctly, it means that every topic will be published with `undefined`, even ones like `successfullyJoined` or `questionCountdown`, etc.

Is there frontend logic, which depends on this mechanism to work?

fafeitsch commented on backend/contest/room.go:
```diff
@@ -169,6 +170,18 @@ func (r *roomImpl) Join(name string) Player {
+		},
+	)
+	delete(r.players, playerKey)
+	// TODO cant return *Player why?
```
`r.players` is a map of strings to _pointers_ to players.

You declare the return type `Player`.

Thus, you have a type mismatch "pointer to player" vs "Player". You can either:
1. have return type `Player`, then you have to `return *player` (the '*' dereferences the pointer)
2. have return type `*Player`, then you can directly `return player`

The `r.players` map uses pointers to make the interface embedding with the `Notifier` work (line 367) – one of Go's many quirks.

I favor the first option because it prevents the player from being changed outside the package (`Leave` is an exported method). Unfortunately, the `FindPlayer` method violates this principle, because one of the method's caller needs the `Notifier` embedding and thus needs a pointer :(.

eduardhuck replied:
Thanks for the detailed explanation.

I also prefer the first solution.

fafeitsch commented:
Thanks for your contribution! What do you think about showing the lave game button also in the waiting room (i.e. when the configuration is done)? Players might want to leave the game while it has not been started yet.

fafeitsch commented on backend/contest/room.go:
```diff
@@ -169,6 +170,18 @@ func (r *roomImpl) Join(name string) Player {
+			p.NotifyPlayerLeft(player.Name, player.Key)
+		},
+	)
+	delete(r.players, playerKey)
```
You could add 

```
if len(r.players) == 0 {
		r.finished = true
}
```


This should delete the room within 5 seconds (the delete room event is logged to cmd).

eduardhuck commented:
> Thanks for your contribution! What do you think about showing the lave game button also in the waiting room (i.e. when the configuration is done)? Players might want to leave the game while it has not been started yet.

Good idea. I will look into it and add the change to this PR.

eduardhuck committed:
feat: handle undefined payload in leaflet

eduardhuck committed:
fix: minor cleanup of return type using existing player

eduardhuck committed:
feat: finish room if no players are left

eduardhuck committed:
chore: remove old TODO and console.log

eduardhuck committed:
feat: add leaveButton to WaitingRoom and refactor into own component

eduardhuck commented:
@fafeitsch I added the `LeaveButton` to the `WaitingRoom` component and extracted it into a component, to avoid redundancy.

I am not sure about the architecture of the frontend and if such a component should import `rpc` or the `store`. In case this contradicts some guidelines, I could add inputs and outputs to the `LeaveButton` if needed.

fafeitsch committed:
fix: remove if-condition for showing the game finished button

fafeitsch commented:
Thanks for contributing. I put the changes onto the main branch.
